### PR TITLE
New Relic Browser Analytics plugin

### PR DESCRIPTION
**Pull request by jmezach**
## Hey, I just made a Pull Request!

As described in #18529 we were looking for a New Relic Browser plugin for our analytics needs on our Backstage instance, but couldn't find one. This adds an implementation of the AnalyticsApi that loads and initialises New Relic Browser's agent if properly configured.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [-] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by benjdlambert** [comment]
@iamEAP do you wanna take a first pass at this?

File: plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (awanlin) [inline plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts:58]: Would it make more sense to change this to be just the endpoint in the config as maybe an enum and then you wouldn't need to this here and below in the `info` section? So instead of `useEuEndpoint` it would be `endpoint` and it would need to be one of `'bam.eu01.nr-data.net' | 'bam.nr-data.net'`

File: plugins/analytics-module-nr/package.json
Comment (awanlin) [inline plugins/analytics-module-nr/package.json:2]: This name needs to be adjusted, it should be `@backstage/plugin-analytics-module-nr` and then that also needs to be updated in your changeset.

I would also like to suggest a larger name change here. I know we have used `ga` for the Google Analytics plugins but I wish we hadn't. I don't think using `nr` is clear and I think having it as `@backstage/plugin-analytics-module-newrelic-browser` would be much better, I'll leave this to @benjdlambert and/or @iamEAP to make the final call on though

**Review by awanlin — CHANGES_REQUESTED** [review]
Thanks for this contribution @jmezach, didn't know this feature existed for New Relic.

I left a few comments for you but overall this is great work! 🚀

File: plugins/analytics-module-nr/package.json
Comment (benjdlambert) [reply plugins/analytics-module-nr/package.json:2]: +1 here, think `newrelic` or `new-relic` should be in the name

File: .github/CODEOWNERS
Comment (jmezach) [reply .github/CODEOWNERS:20]: Yeah, I noticed that too. Probably some kind of whitespace issue. I've added @backstage/maintainers to it as well.

File: .changeset/rude-feet-sparkle.md
Comment (jmezach) [reply .changeset/rude-feet-sparkle.md:5]: Oh yeah, that is probably useful for adopters. Thanks.

File: plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (jmezach) [reply plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts:58]: I chose this as I'm not entirely sure that users would be familiar with those particular endpoints. They are documented of course on the New Relic side, but I think users would have an easier time to reason about either the US or the EU data centers, instead of specific endpoints. But I'd be happy to change it.

File: plugins/analytics-module-nr/package.json
Comment (jmezach) [reply plugins/analytics-module-nr/package.json:2]: Yeah, I thought about that too. I'll change it to `@backstage/plugin-analytics-module-newrelic-browser`. Should I change this plugins folder name to `analytics-module-newrelic-browser` as well?

**Comment by benjdlambert** [comment]
Woops - I think that this will need a rebase from the main branch to get this to look nice again, looks like a load of other commits have just popped in here!

File: plugins/analytics-module-nr/package.json
Comment (benjdlambert) [reply plugins/analytics-module-nr/package.json]: Yep I would change the folder name to reflect the package name too.

**Comment by jmezach** [comment]
@benjdlambert I actually did a rebase just now because of the conflicts on the `yarn.lock` file. Not sure what happened here, but I'll look into it.

**Comment by jmezach** [comment]
@benjdlambert I just did another rebase and I think it looks ok now again. Perhaps this was caused by the issue with PR's on GitHub earlier this morning.

File: plugins/analytics-module-nr/package.json
Comment (jmezach) [reply plugins/analytics-module-nr/package.json]: Made that change too.

File: plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (jmezach) [reply plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts]: @awanlin @benjdlambert I guess this is the only remaining issue here. Any thoughts on this?

File: plugins/analytics-module-new-relic/package.json
Comment (iamEAP) [inline plugins/analytics-module-new-relic/package.json:14]: Don't believe this is the correct role.

I'm on mobile, so I don't have the correct value handy, but something like frontend module...

File: plugins/analytics-module-new-relic/README.md
Comment (iamEAP) [inline plugins/analytics-module-new-relic/README.md:35]: Would it be Worth highlighting the optionality of this? I assume the reason you made it possible to omit is for privacy reasons.

File: plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (iamEAP) [inline plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts:112]: It's been a minute since I used NR, but I recall that they have a way of differentiating page views from page actions.

My recommendation would be to hook onto events with action "navigate" and send them along as page views, and then let the remaining actions pass as page actions.

File: plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (iamEAP) [inline plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts:110]: Important thing to note about this `rest` param is that it's not a flat object. The keys `context` and `attributes` will contain a lot of extra metadata that you will almost certainly want to decorate your events with.

If I recall correctly, NR only supports flat key/value pairs. So you may want to pull those out and merge them in.

You may even want/need to namespace the keys if each (so that context keys and attributes keys that are the same don't clobber each other).

**Review by iamEAP — COMMENTED** [review]
Hey! Thanks for the contribution. I've been waiting for someone to contribute something like this; lots of amazing possibilities with NRQL.

Beyond the comments I've left, I'd also encourage you to add this to the table of supported analytics api integrations on the plugin analytics page in the docs!

File: plugins/analytics-module-new-relic/package.json
Comment (jmezach) [reply plugins/analytics-module-new-relic/package.json:14]: Changed it to frontend-plugin-module as per [this doc](https://backstage.io/docs/local-dev/cli-build-system/#package-roles).

File: plugins/analytics-module-new-relic/README.md
Comment (jmezach) [reply plugins/analytics-module-new-relic/README.md:35]: Good point. I've added a section about User ID's to the README. I noticed that the Google Analytics implementation also calculates a hash of the user entity reference by default. Perhaps we should add that here as well?

Comment (iamEAP) [reply plugins/analytics-module-new-relic/README.md:35]: The GA implementation was following Google guidance on minimizing PII (a terms of service issue); you could take a similar approach if it makes sense re: NR's terms of service.

File: plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (awanlin) [reply plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts]: Still think it would look much nicer in code if you do something like this: 

`endpoint?: 'bam.eu01.nr-data.net' | 'bam.nr-data.net';`

An example: https://github.com/backstage/backstage/blob/c4a9eb45ab17668c3f04ac313f99d6430a7349f6/plugins/analytics-module-ga4/config.d.ts#L47

You could list the endpoints in the `README` and include a link to the relevant documentation. Also, maybe making this visible might help those with tight networking policies identify this as something they might need to open up? 

What do you think @benjdlambert ?

Comment (benjdlambert) [reply plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts]: Think this makes sense @awanlin :tada:

Comment (jmezach) [reply plugins/analytics-module-nr/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts]: Yeah, that makes sense. I've made these changes just now.

File: plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts
Comment (jmezach) [reply plugins/analytics-module-new-relic/src/apis/implementations/AnalyticsApi/NewRelicBrowser.ts:110]: I'm also attaching the custom attributes now.

File: plugins/analytics-module-new-relic/README.md
Comment (jmezach) [reply plugins/analytics-module-new-relic/README.md]: I couldn't really find any guidance on this in the New Relic docs, so I added a similar implementation as the GA one. I guess folks can always just return the original user ID if they want to, but then at least it is an opt-in rather than default behaviour.

**Comment by jmezach** [comment]
I think I've processed all the comments. @benjdlambert @iamEAP @awanlin Could you have another look at this?

File: .github/CODEOWNERS
Comment (awanlin) [inline .github/CODEOWNERS:20]: This needs an update, please

File: plugins/analytics-module-newrelic-browser/config.d.ts
Comment (awanlin) [inline plugins/analytics-module-newrelic-browser/config.d.ts:22]: This probably needs to be updated?

File: plugins/analytics-module-newrelic-browser/README.md
Comment (awanlin) [inline plugins/analytics-module-newrelic-browser/README.md:13]: This will need to be updated with the new name

**Review by awanlin — CHANGES_REQUESTED** [review]
Looking really good @jmezach! Just left a few clean up comments

File: .github/CODEOWNERS
Comment (jmezach) [reply .github/CODEOWNERS:20]: Good catch, thanks :).

File: plugins/analytics-module-newrelic-browser/config.d.ts
Comment (jmezach) [reply plugins/analytics-module-newrelic-browser/config.d.ts:22]: Ah yes, of course.

File: plugins/analytics-module-newrelic-browser/README.md
Comment (jmezach) [reply plugins/analytics-module-newrelic-browser/README.md:13]: Sure thing.

**Comment by jmezach** [comment]
@benjdlambert @iamEAP @awanlin I think we're all good here?

**Comment by jmezach** [comment]
Looks like TechDocs pipeline is failing, not entirely sure why though.

**Review by awanlin — APPROVED** [review]

**Comment by awanlin** [comment]
> Looks like TechDocs pipeline is failing, not entirely sure why though.

I kicked off the build again to see if it was a one time thing, getting further so far

**Comment by jmezach** [comment]
@benjdlambert Can this be merged?

**Comment by jmezach** [comment]
@backstage-service I think this can be merged?

File: plugins/analytics-module-newrelic-browser/config.d.ts
Comment (benjdlambert) [inline plugins/analytics-module-newrelic-browser/config.d.ts:20]: should we change this to `new-relic` or `newRelic`?

**Review by benjdlambert — CHANGES_REQUESTED** [review]
Sorry for the delay in this PR - just a quick comment about the config schema.

**Comment by jmezach** [comment]
@benjdlambert No problem, I know things can get busy. Just wanted to make sure this didn't fall through the cracks.

File: plugins/analytics-module-newrelic-browser/config.d.ts
Comment (benjdlambert) [reply plugins/analytics-module-newrelic-browser/config.d.ts:20]: yeah think it's nicer to camel case things :)

Comment (jmezach) [reply plugins/analytics-module-newrelic-browser/config.d.ts]: Done :).

**Comment by jmezach** [comment]
@benjdlambert Microsite build is taking an awful lot of time, not sure what's going on there

**Comment by jmezach** [comment]
@benjdlambert I think we're good to go now?

**Review by benjdlambert — APPROVED** [review]
Thanks - let's go with this! :tada:

**Pull request merged**
